via/views/error.py

```diff
@@ -0,0 +1,96 @@
+"""Error views to handle when things go wrong in the app."""
```

Comment:
Should be `exceptions.py` (even though that's not consistently done in other apps yet): https://stackoverflow.com/c/hypothesis/questions/295
Pyramid doesn't have "error views" it has [exception views](https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/views.html?highlight=%22exception%20views%22#custom-exception-views) (they can be used for any kind of exception, not just ones that represent errors)

Reply:
I'm confused on our usage of the words error and exception here. I think you have a definition in mind and it would be good to see it written down.

When we've discussed errors vs. exceptions before, we've consistently described everything as an error in the class name: "ExampleError", but kept them all in a file called "exceptions.py". From our previous descriptions I couldn't really find anything you'd describe as an exception.

For my experience in Java, which has a fairly strong interpretation there are two very distinct things:

 * Exceptions - Happen in normal flow, and you should catch them
 * Errors - Extremely serious, no point in trying to catch it (CPU has gone away, all memory corrupted etc.)

I can't find any similarly concrete definition for Python. It seems from the basic library that the difference is mostly that "Exception" and "BaseException" are mostly base classes, and all concrete classes are "<Something>Error". So all errors are exceptions, and all custom classes should be errors. 

So as far as I can tell, we shouldn't be declaring "exceptions" as such, rather we should be declaring "errors". So "exceptions.py" is an accurate name to give the file we declare them in, but in the same way that "objects.py" is as well.

I have no objection to renaming the file by the way, but it would be good to understand why, and check that there is a sensible rule at play here.

Reply:
I think it's similar in Python to in Java, but Python does not have the concept that errors should not be caught:

* An exception is just an interruption-to-the-normal-control-flow thing, and not necessarily an error. E.g. `GeneratorExit`, `KeyboardInterrupt`, `StopIteration`, `StopAsyncIteration`, `SystemExit`.

* Errors are a subclass of exceptions and represent a "something went wrong", and their name ends in `Error`. E.g. `ImportError`, `AttributeError`, there are many.

Not all concrete exception classes in the stdlib are `Error`.

In Pyramid it's very clear that there's no such thing as an "error view". They're called "exception views", and they don't necessarily represent errors. By default you can raise an HTTP _exception_ (from `pyramid.httpexceptions`) and there are default exception views that will catch these. Some HTTP exceptions are arguably error-ish (unauthorized, not found, ...) (and only arguably: it's not `HTTPBadRequest` not `HTTPBadRequestError`, the only one that's actually called an error is `HTTPInternalServerError`) but some HTTP exceptions clearly aren't errors (redirects, you can even raise `HTTPCreated` or `HTTPAccepred` etc). HTTP exceptions and exception views are a control-flow thing in Pyramid not just for errors.

Similarly when you write your own custom exception views, the custom exception classes you're catching can be error-like exceptions or not.

This terminology is consistent in Pyramid so we should just follow it.

Reply:
I'm just trying to work out the definition, rather than the terminology.

So "exceptions" that are not "errors" are really just those that are a part of normal execution flow in completely non error situations. I.e. coming to the end of an iterator is an expected part of how things work when their going right.

So what changes an "exception" into being an "error" is that (annoyingly) something exceptional happened which isn't expected.

To map to the Java concepts:

 * Python exception = for loops, control flow etc. (Java doesn't use exceptions like this I don't think)
 * Python error = Java exception (it's bad, but you should catch it)
 * Python SyntaxError maybe? = Java errors (looks like this concept might be missing from Python mostly)

Reply:
For the sake of brevity I'm just going to go with this naming scheme, but... If "exceptions" are exclusively normal things to happen, then we'll never show an error page as a result of one right? 100% of what we show is under this definition an "error"?

Reply:
No, the intention of `views/exception.py` is that it's where we'd put all that directories custom exception views, including both error-like and non-error-like exception views should it have any. We've discussed this previously, e.g. https://github.com/hypothesis/lms/issues/1245 and https://stackoverflow.com/c/hypothesis/questions/295.

It's honestly just a misnomer to refer to what the Pyramid docs and code always call "exception views" (`@exception_view_config()`, `add_exception_view()`, ...) as "error views" and means if we ever want to add a non-error-like exception view we'd have to put it elsewhere.

Reply:
From what I can see from the hierarchy there is a distinction in a way:

https://docs.python.org/3/library/exceptions.html

* `SystemExit`, `KeyboardInterrupt` and `GeneratorExit` are direct children of `BaseException`
* Everything else is a child of `Exception`, including `StopIteration`
* All warnings are children of `Warning` in addition

Not sure what to make of that. I'm sure there's a reason that `StopIteration` is an odd sausage here, but I can't think of it. It would have been nice if the root was "Raisable" (or something non shit) and then "Exception" was only one branch.

I think the distinction of an "error" vs an "exception" might be one we've invented. The stdlib docs don't mention any distrinction and use the terms interchangeably:

> The built-in **exceptions** listed below can be generated by the interpreter or built-in functions. Except where mentioned, they have an “associated value” indicating the detailed cause of the **error**.

Reply:
I think the broader question of whether there are separate concepts of "exception" and "error" in Python generally is a distraction here. To me there's a fairly clear semantic distinction between error and non-error exceptions that can be seen in both the stdlib's exceptions and in Pyramid's HTTP exceptions (where you have an `HTTPException` base class that the 2xx and 3xx except subclass directly, with an `HTTPError` subclass that the 4xx and 5xx exceptions subclass). But putting that aside, they're just called exception views in Pyramid not error views. This seems pretty incontrovertible if you look at the Pyramid API and docs. I don't think there's a good reason for us to decide to call them error views instead?

Reply:
Anywho, as long as we have our own definition I suppose that's enough.

Maybe something like:

**Style:** We define errors (as a sub-class of exceptions) as being those classes raised when something is wrong (e.g. `ValueError`, `requests.exception.InvalidURL`). While all errors are exceptions, some exceptions are not errors, like `StopIteration` and are a part of normal part of control flow.

**Style:** When defining custom errors and exceptions they are stored in a file called `exceptions.py` even if the file happens to contain only errors.

**Style:** Similarly even when Pyramid exception views are used exclusively to catch errors (rather than pure exceptions like `HTTPFound`) they will in a file called `exceptions.py`. This matches the terminolog in Pyramid and allows the possibility of catching and reacting to exceptions like `HTTPFound`, for example to modify them in some way.

Reply:
I've added a longer version of this as a S/O question: https://stackoverflow.com/c/hypothesis/questions/295

Also renamed everything I could find to be more exception related.